src/ripple/overlay/impl/PeerSet.cpp

```diff
@@ -24,32 +24,53 @@
 
 namespace ripple {
 
-using namespace std::chrono_literals;
-
-PeerSet::PeerSet(
-    Application& app,
-    uint256 const& hash,
-    std::chrono::milliseconds interval,
-    beast::Journal journal)
-    : app_(app)
-    , m_journal(journal)
-    , mHash(hash)
-    , mTimeouts(0)
-    , mComplete(false)
-    , mFailed(false)
-    , mProgress(false)
-    , mTimerInterval(interval)
-    , mTimer(app_.getIOService())
+class PeerSetImpl : public PeerSet
+{
+public:
+    using ScopedLockType = std::unique_lock<std::recursive_mutex>;
+
+    PeerSetImpl(Application& app);
+
+    void
+    addPeers(
+        std::size_t limit,
+        std::function<bool(std::shared_ptr<Peer> const&)> hasItem,
+        std::function<void(std::shared_ptr<Peer> const&)> onPeerAdded) override;
+
+    /** Send a GetLedger message to one or all peers. */
+    void
+    sendRequest(
+        ::google::protobuf::Message const& message,
+        protocol::MessageType type,
+        std::shared_ptr<Peer> const& peer) override;
+
+    void
+    visitAddedPeers(std::function<void(Peer::id_t)> onPeer) override;
+
+    std::size_t
+    countAddedPeers() override;
+
+private:
+    // Used in this class for access to boost::asio::io_service and
+    // ripple::Overlay.
+    Application& app_;
+    beast::Journal m_journal;
+    mutable std::recursive_mutex mLock;
+
+    /** The identifiers of the peers we are tracking. */
+    std::set<Peer::id_t> mPeers;
```

Comment:
I have the same question. This is from existing code. For the new usages, ie. `LedgerDeltaAcquire` and `SkipListAcquire`, it doesn't matter much since I expect very a small number of items (e.g. 2). For the existing usages, I think it will be a good change. But I don't feel I know enough.